### ISSUE TITLE
Graceful shutdown for persistent connections

### ISSUE THREAD

**Issue by zhu-he** [actix/actix-web#3169]
When an `actix-web` server receives a SIGTERM signal, it will stop accepting new connections and wait for all active connections to close. After the shutdown timeout configured by `HttpServer::shutdown_timeout`, the server will force close all active connections.

In our case, we have several servers behind a layer 4 load balancer. When we want to deploy a new version of our service, we will perform a rolling update. We will stop one old server, deploy the new version of the service, and start the new server. Before the old server is stopped, the load balancer will stop making new connections to the old server. However, the old server will still have some persistent connections. If some request is received before the old server is stopped, but not finished before the shutdown timeout, the client will receive a `connection reset by peer` error.

After some research, we found that other web frameworks have a more graceful shutdown strategy. For HTTP/1.1, the server will add a `Connection: close` header to all subsequent responses. For HTTP/2, the server will send a `GOAWAY` frame to the client. This will allow the client to close the connection gracefully.

For example, in the `hyper` crate, the above strategy is implemented in the [code](https://github.com/hyperium/hyper/blob/v0.14.27/src/server/conn.rs#L744-L761). All `hyper`-based web frameworks, such as `axum`, can benefit from this graceful shutdown strategy.

We have tried to implement this graceful shutdown strategy in `actix-web` using middleware. However, some difficulties are encountered. For example, the middleware cannot access the shutdown signal, but luckily signals are the only shutdown condition in our case, and we can directly watch them. Another difficulty is that the middleware can only solve the problem in HTTP/1.1, but not in HTTP/2, as the middleware cannot access the underlying connection.

We hope that `actix-web` can implement this graceful shutdown strategy. This will make `actix-web` more suitable for production environments.

**Comment by liketoeatcheese** [comment]
Hi @zhu-he ,

I'm using version 4.5.1 of Actix-web, and when I do `kill -SIGTERM {actix server pid}`, I got graceful shutdown logs:

```log
[2024-03-01T15:31:06.361604665Z]  INFO: testapp/2721791 on archlinux: starting 32 workers (target=actix_server::builder, line=240)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/builder.rs
[2024-03-01T15:31:06.361654984Z]  INFO: testapp/2721791 on archlinux: Actix runtime found; starting in Actix runtime (target=actix_server::server, line=196)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/server.rs
[2024-03-01T15:49:20.526868842Z]  INFO: testapp/2721791 on archlinux: SIGTERM received; starting graceful shutdown (target=actix_server::server, line=328)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/server.rs
[2024-03-01T15:49:20.526993052Z]  INFO: testapp/2721791 on archlinux: accept thread stopped (target=actix_server::accept, line=143)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/accept.rs
[2024-03-01T15:49:20.527000752Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
[2024-03-01T15:49:20.526999672Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
[2024-03-01T15:49:20.526999202Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
[2024-03-01T15:49:20.527008162Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
[2024-03-01T15:49:20.527039163Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
[2024-03-01T15:49:20.527040443Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
[2024-03-01T15:49:20.527045863Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
[2024-03-01T15:49:20.527062153Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
[2024-03-01T15:49:20.527048713Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
[2024-03-01T15:49:20.527065093Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
[2024-03-01T15:49:20.527065673Z]  INFO: testapp/2721791 on archlinux: shutting down idle worker (target=actix_server::worker, line=595)
    file: /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/worker.rs
```

Not sure which version you are using, maybe the maintainers did add this on the later version

**Comment by johanatan** [comment]
This is likely mislabeled if it in fact is already implemented. There's nothing easier than 'nothing'.

**Comment by Sytten** [comment]
No this is clearly not implemented, we are having trouble with rolling updates in our cloud provider too. They send a SIGTERM to the service but still send some HTTP requests to the service, so when actix kills the connections they return a 502 to our clients. We don't see this with another service that is based on hyper.

**Comment by Sytten** [comment]
Hey @robjtede I just spend the whole day debugging graceful shutdowns in our production service.
I think I found at least one bug in actix (more on that later), though I have no idea how to fix it.

**For context:** We deploy our service on render.com in a scratch container. Recently we observed that the render load balancer (LB) started returning 502 during deployments. This happens when the LB tries to send a request to an upstream server and fails. This doesn't happen with our other server that is based on poem/hyper.

**TLDR of the problem:** The LB of render doesn't handle well when connections are killed by the server, you must keep the connections alive even when you receive a SIGTERM and wait for the LB to close the connection. You can send a `Connection: close` on the next Http request, but you must still keep it alive.

It works well in hyper because of two elements:
- Poem/Hyper will start sending `Connection: close` when the server is shutdown
- Poem/Hyper will continue to poll the connection indefinitely (unless you specify a timeout), it will **NOT** initiate the TCP shutdown (aka it won't drop the `TCPStream`).

So we first thing we did was to introduce a middleware like @zhu-he mentioned to add the connection close. Here is the source if someone needs it.

<details>

<summary>Middleware source code</summary>

```rust 
use std::sync::atomic::{AtomicBool, Ordering};

use actix_http::header::{CONNECTION, HeaderValue};
use actix_web::Error;
use actix_web::dev::{Service, ServiceRequest, ServiceResponse, Transform};
use futures::FutureExt;
use futures::future::{LocalBoxFuture, Ready, ok};

static SHUTDOWN: AtomicBool = AtomicBool::new(false);

// Required for graceful shutdown until https://github.com/actix/actix-web/issues/3169
pub struct ConnectionClose;

impl ConnectionClose {
    pub fn shutdown() {
        SHUTDOWN.store(true, Ordering::Relaxed);
    }
}

impl<S> Transform<S, ServiceRequest> for ConnectionClose
where
    S: Service<ServiceRequest, Response = ServiceResponse, Error = Error> + 'static,
{
    type Response = ServiceResponse;
    type Error = Error;
    type Transform = ConnectionCloseService<S>;
    type InitError = ();
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ok(ConnectionCloseService { service })
    }
}

pub struct ConnectionCloseService<S> {
    service: S,
}

impl<S> Service<ServiceRequest> for ConnectionCloseService<S>
where
    S: Service<ServiceRequest, Response = ServiceResponse, Error = Error> + 'static,
{
    type Response = ServiceResponse;
    type Error = Error;
    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;

    actix_service::forward_ready!(service);

    fn call(&self, req: ServiceRequest) -> Self::Future {
        let fut = self.service.call(req);
        async move {
            let mut resp = fut.await?;
            if SHUTDOWN.load(Ordering::Relaxed) {
                resp.headers_mut()
                    .insert(CONNECTION, HeaderValue::from_static("close"));
                return Ok(resp);
            }
            Ok(resp)
        }
        .boxed_local()
    }
}
```

</details>

We then disabled signal handling in actix and made our own (based on the same code as actix).
This didn't solve the problem, but it is a step in the right direction.

<details>

<summary>Signal handling future</summary>

```rust
use std::fmt;
use std::future::Future;
use std::pin::Pin;
use std::task::{Context, Poll};

const TARGET: &str = "cloud|signal";

#[derive(Debug, Clone, Copy, PartialEq)]
#[allow(dead_code)]
pub enum SignalKind {
    Interrupt,
    Terminate,
    Quit,
}

impl SignalKind {
    pub fn graceful(&self) -> bool {
        match self {
            SignalKind::Terminate => true,
            SignalKind::Interrupt | SignalKind::Quit => false,
        }
    }
}

impl fmt::Display for SignalKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            SignalKind::Interrupt => "SIGINT",
            SignalKind::Terminate => "SIGTERM",
            SignalKind::Quit => "SIGQUIT",
        })
    }
}

pub struct Signals {
    #[cfg(not(unix))]
    signals: futures::future::BoxFuture<'static, std::io::Result<()>>,

    #[cfg(unix)]
    signals: Vec<(SignalKind, tokio::signal::unix::Signal)>,
}

impl Signals {
    pub fn new() -> Self {
        log::info!(target: TARGET, "setting up OS signal listener");

        #[cfg(not(unix))]
        {
            Signals {
                signals: Box::pin(tokio::signal::ctrl_c()),
            }
        }

        #[cfg(unix)]
        {
            use tokio::signal::unix;

            let sig_map = [
                (unix::SignalKind::interrupt(), SignalKind::Interrupt),
                (unix::SignalKind::terminate(), SignalKind::Terminate),
                (unix::SignalKind::quit(), SignalKind::Quit),
            ];

            let signals = sig_map
                .iter()
                .filter_map(|(kind, sig)| {
                    unix::signal(*kind)
                        .map(|tokio_sig| (*sig, tokio_sig))
                        .map_err(|e| {
                            log::error!(
                                target: TARGET,
                                "cant initialize stream handler for {sig}: {e:?}",
                            )
                        })
                        .ok()
                })
                .collect::<Vec<_>>();

            Signals { signals }
        }
    }
}

impl Future for Signals {
    type Output = SignalKind;

    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        #[cfg(not(unix))]
        {
            self.signals
                .as_mut()
                .poll(cx)
                .map(|_| SignalKind::Interrupt)
        }

        #[cfg(unix)]
        {
            for (sig, fut) in self.signals.iter_mut() {
                if fut.poll_recv(cx).is_ready() {
                    log::info!(target: TARGET, "{sig} received");
                    return Poll::Ready(*sig);
                }
            }

            Poll::Pending
        }
    }
}
```

</details>


```rust
// Run server
    let signals = Signals::new();
    let server = server_builder.run().await?;
    let handle = server.handle();
    tokio::spawn(async move {
        let signal = signals.await;
        cloud_middlewares::ConnectionClose::shutdown();
        handle.stop(signal.graceful()).await;
    });
    server.await?;
```

We then tried to reproduce the behaviour of poem/hyper by disabling or making very long all the timeouts.
We tried a bunch of combinations, but ended with:

```rust
let server = HttpServer::new(move || {
            App::new()
                .wrap(ConnectionClose)
                .configure(yourservices)
        })
        .keep_alive(KeepAlive::Os) // This effectively disables the actix timer for keep-alive
        .client_disconnect_timeout(Duration::ZERO) // This disables the client timeout when shutting down the connection
        .disable_signals()
        .workers(10)
        .bind(self.addr)?
        .run();
```

This didn't work either, which I find very strange. I saw in the logs things like `graceful worker shutdown; finishing 2 connections` but then 1s after the worker stopped and the connection was killed. I looked in the code and I saw that the 1s is an [hard coded value](https://github.com/actix/actix-net/blob/c9cea8644543dfdcc7da29a31c9157ae58b1088a/actix-server/src/worker.rs#L661-L662) in the worker. I think there is a bug there because it will **NOT** wait for the connection to finish on it's own, at least I could not make it work. Also the `shutdown_timeout` didn't seem to have any impact.

So what made it work for us was to introduce an artificial sleep when the signal is received. We start returning `Connection: close` but we keep actix alive for 30s. We also bumped the default render.com `maxShutdownDelaySeconds` to 60s.

```rust
tokio::spawn(async move {
        let signal = signals.await;
        cloud_middlewares::ConnectionClose::shutdown();
        tokio::time::sleep(Duration::from_secs(30)).await; // This is the patch
        handle1.stop(signal.graceful()).await;
    });
```

So now the behaviour is:
1. Render sends a SIGTERM
2. We start returning `Connection: close`, but keep the system running
3. (30s later) We start a graceful shutdown
4. Worst case the shutdown_timeout will timeout after 30s
5. Worst worst case render sends a SIGKILL after 60s

So far this allowed us to have a proper graceful shutdown without any downtime. Hopefully this will help others facing a similar situation and provide so help to you @robjtede if you want to work on the problem.

There is definitely some fault on render here, their LB really s*cks but still it's better for actix to be as reliable as possible.

**Comment by robjtede** [comment]
I've figured this can be done without breaking changes. Fix incoming.
